diff --git a/src/functions.cpp b/src/functions.cpp
index 469d911..ae8372c 100644
--- a/src/functions.cpp
+++ b/src/functions.cpp
@@ -19,6 +19,7 @@
 #include <signal.h>
 #include <sys/wait.h>
 #include <spawn.h>
+#include <sys/utsname.h>
 
 #include "devices.h"
 #include "fbink.h"
@@ -347,8 +348,10 @@ void prepareVariables() {
   // Handle turning off the LED usage option
   ledManagerAccurate();
 
-  if(normalContains(executeCommand("uname -r | grep -o '[^-]*$'"), "n306c")) {
-    log("It is the nia model c", emitter);
+  struct utsname uname_data;
+  uname(&uname_data);
+  if(normalContains(uname_data.release, "n306c")) {
+    log("Running on n306c", emitter);
     isNiaModelC = true;
   }
 }
