diff --git a/etc/init.d/usbnet b/etc/init.d/usbnet
index 775b1f3..d1c908a 100755
--- a/etc/init.d/usbnet
+++ b/etc/init.d/usbnet
@@ -14,6 +14,7 @@ start() {
 	HOST_ADDRESS="$(cat /boot/flags/USBNET_HOST_ADDRESS 2>/dev/null)"
 	IP="$(cat /boot/flags/USBNET_IP 2>/dev/null)"
 	DEVICE="$(cat /opt/inkbox_device)"
+	[ "$(uname -r | grep -o '[^-]*$')" == "n306c" ] && DEVICE_VARIANT="n306c"
 
 	if [ "${ROOTED}" == "true" ]; then
 		if [ "${FLAG}" == "true" ]; then
@@ -21,13 +22,15 @@ start() {
 			if [ "${DEVICE}" == "n705" ] || [ "${DEVICE}" == "n905b" ] || [ "${DEVICE}" == "n905c" ] || [ "${DEVICE}" == "n613" ] || [ "${DEVICE}" == "n236" ] || [ "${DEVICE}" == "n437" ]; then
 				insmod "/modules/g_ether.ko" dev_addr="${DEVICE_ADDRESS}" host_addr="${HOST_ADDRESS}"
 			elif [ "${DEVICE}" == "n306" ] || [ "${DEVICE}" == "n873" ] || [ "${DEVICE}" == "bpi" ]; then
-				insmod "/modules/fs/configfs/configfs.ko"
-				insmod "/modules/drivers/usb/gadget/libcomposite.ko"
-				insmod "/modules/drivers/usb/gadget/function/u_ether.ko"
-				insmod "/modules/drivers/usb/gadget/function/usb_f_ecm.ko"
-				[ -e "/modules/drivers/usb/gadget/function/usb_f_eem.ko" ] && insmod "/modules/drivers/usb/gadget/function/usb_f_eem.ko"
-				insmod "/modules/drivers/usb/gadget/function/usb_f_ecm_subset.ko"
-				insmod "/modules/drivers/usb/gadget/function/usb_f_rndis.ko"
+				if [ "${DEVICE_VARIANT}" != "n306c" ]; then
+					insmod "/modules/fs/configfs/configfs.ko"
+					insmod "/modules/drivers/usb/gadget/libcomposite.ko"
+					insmod "/modules/drivers/usb/gadget/function/u_ether.ko"
+					insmod "/modules/drivers/usb/gadget/function/usb_f_ecm.ko"
+					[ -e "/modules/drivers/usb/gadget/function/usb_f_eem.ko" ] && insmod "/modules/drivers/usb/gadget/function/usb_f_eem.ko"
+					insmod "/modules/drivers/usb/gadget/function/usb_f_ecm_subset.ko"
+					insmod "/modules/drivers/usb/gadget/function/usb_f_rndis.ko"
+				fi
 				insmod "/modules/drivers/usb/gadget/legacy/g_ether.ko" dev_addr="${DEVICE_ADDRESS}" host_addr="${HOST_ADDRESS}"
 			elif [ "${DEVICE}" == "n249" ] || [ "${DEVICE}" == "kt" ]; then
 				modprobe g_ether dev_addr="${DEVICE_ADDRESS}" host_addr="${HOST_ADDRESS}"
diff --git a/sbin/ipd b/sbin/ipd
index 9f37efd..2e247d1 100755
Binary files a/sbin/ipd and b/sbin/ipd differ
